### PR TITLE
feat(exports): add chunked job pipeline for large audience exports

### PR DESCRIPTION
**Pull request by FusionBrah**
Issue: #2092

# Description

## Problem

Large audience exports (followers/subscribers CSV) timeout because the entire export happens in a single Sidekiq job. Users with large audiences never receive their export emails.

## Solution

Implement a chunked export pipeline (following the existing `SalesExport` pattern):

- **CreateAndEnqueueChunksJob** - splits audience into 1000-member chunks
- **ProcessChunkJob** - processes each chunk independently with `lock: :until_executed`
- **CompileChunksJob** - assembles final CSV and sends email

Small audiences (≤1000 members) continue using the existing synchronous path for efficiency.

**Additional fix:** Added `validate_options!` to `CreateAndEnqueueChunksJob` to ensure consistent validation between sync and async paths. Without this, passing all-false options to the async path would silently export ALL members instead of failing.

**Depends on:** #2532 (database models)

---

# Before/After

Backend-only change

---

# Test Results

Screenshot 2026-01-02 at 10 04 26 pm

---

# Checklist

- [x] I have read the contributing guidelines
- [x] I have watched Gumroad PR review livestreams
- [x] I have performed a self-review and left review comments on my PR
- [x] I have added/updated tests for my changes

---

# AI Disclosure

AI (Claude Code) was used for:
- Code implementation (Sidekiq job classes, chunking logic, CSV compilation)
- Test writing (RSpec specs including edge cases and race condition tests)
- Debugging test failures related to ordering expectations
- Performance testing with 5,500 member dataset
- Code review identifying validation gap between sync/async paths

### CODE REVIEW

**Comment by FusionBrah** [comment]
Local test results:

All 38 tests passing locally.

CheckPassed 1
CheckPassed2

File: app/sidekiq/exports/audience/compile_chunks_job.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+require "csv"
+
+class Exports::Audience::CompileChunksJob
+  include Sidekiq::Job
+  sidekiq_options retry: 5, queue: :low, lock: :until_executed
```
Comment (FusionBrah) [inline app/sidekiq/exports/audience/compile_chunks_job.rb:7]: `lock: :until_executed` ensures the compile job only runs once even if triggered by multiple chunks finishing simultaneously. Following the idempotency pattern from CLAUDE.md guidelines.

File: app/sidekiq/exports/audience/create_and_enqueue_chunks_job.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  def perform(export_id)
+    @export = AudienceExport.find(export_id)
+    create_chunks
+    enqueue_chunks
+  end
+
+  private
+    def create_chunks
+      @export.chunks.in_batches(of: 100).delete_all
```
Comment (FusionBrah) [inline app/sidekiq/exports/audience/create_and_enqueue_chunks_job.rb:18]: Clearing stale chunks first ensures idempotency if this job retries after a partial failure.

File: app/sidekiq/exports/audience/create_and_enqueue_chunks_job.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+      member_ids = build_query.order(:id).pluck(:id)
+
+      if member_ids.empty?
```
Comment (FusionBrah) [inline app/sidekiq/exports/audience/create_and_enqueue_chunks_job.rb:22]: Edge case: if the audience is empty (e.g., all members were deleted between enqueue and execution), skip straight to compile to send an empty CSV rather than failing.

File: app/sidekiq/exports/audience/process_chunk_job.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+class Exports::Audience::ProcessChunkJob
+  include Sidekiq::Job
+  sidekiq_options retry: 5, queue: :low, lock: :until_executed
```
Comment (FusionBrah) [inline app/sidekiq/exports/audience/process_chunk_job.rb:5]: `lock: :until_executed` prevents duplicate processing if the same chunk job is enqueued multiple times during retries. Following the Sidekiq guidelines from CLAUDE.md.

File: app/sidekiq/exports/audience/process_chunk_job.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      members = AudienceMember.where(id: @chunk.member_ids)
+                              .select(:email, :min_created_at)
+                              .order(:min_created_at)
+
+      members_data = members.map { |m| [m.email, m.min_created_at&.to_s] }
+
+      @chunk.update!(members_data: members_data, processed: true)
+    end
+
+    def trigger_compile_if_complete
```
Comment (FusionBrah) [inline app/sidekiq/exports/audience/process_chunk_job.rb:26]: Only the last chunk to finish triggers the compile job. This check is safe even with parallel processing since we're checking the database state atomically.

**Comment by FusionBrah** [comment]
## Additional Fix: Validation for Async Export Path

During code review, I identified an inconsistency between the sync and async export paths:

| Path | Behavior with all-false options |
|------|--------------------------------|
| Sync (≤1000) | Raises `ArgumentError` via `AudienceExportService#validate_options!` |
| Async (>1000) | Would silently export ALL members (no validation) |

**Fix Applied (235d1bd5):**

Added `validate_options!` to `CreateAndEnqueueChunksJob` to ensure consistent behavior:

```ruby
def validate_options!
  options = @export.audience_options.with_indifferent_access
  return if options[:followers] || options[:customers] || options[:affiliates]

  raise ArgumentError, "At least one audience type (followers, customers, or affiliates) must be selected"
end
```

This ensures the async path now fails fast with a clear error instead of silently exporting all members when invalid options are passed.

**Test added:** Verifies error is raised when all audience types are false.

All 39 tests pass ✓

File: app/sidekiq/exports/audience/create_and_enqueue_chunks_job.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      query = @export.seller.audience_members.select(:id)
+      conditions = []
+      options = @export.audience_options.with_indifferent_access
+      conditions << "follower = true" if options[:followers]
+      conditions << "customer = true" if options[:customers]
+      conditions << "affiliate = true" if options[:affiliates]
+      query.where(conditions.join(" OR "))
+    end
+
+    def validate_options!
```
Comment (FusionBrah) [inline app/sidekiq/exports/audience/create_and_enqueue_chunks_job.rb:47]: Added validation to match `AudienceExportService#validate_options!` behavior. This ensures the async path fails fast with a clear error if invalid options are passed, rather than silently exporting all members via an empty `WHERE` clause.

File: spec/sidekiq/exports/audience/create_and_enqueue_chunks_job_spec.rb [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+      it "triggers CompileChunksJob immediately" do
+        described_class.new.perform(export.id)
+
+        expect(export.chunks.count).to eq(0)
+        expect(Exports::Audience::CompileChunksJob).to have_enqueued_sidekiq_job(export.id)
+      end
+    end
+
+    context "with invalid options" do
+      it "raises error when all audience types are false" do
```
Comment (FusionBrah) [inline spec/sidekiq/exports/audience/create_and_enqueue_chunks_job_spec.rb:60]: Only testing the 'all false' case because empty `{}` options are already blocked by the `AudienceExport` model validation (`audience_options can't be blank`).

File: app/models/audience_export.rb [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+      candidate = SecureRandom.alphanumeric(12).downcase
+
+      while exists?(external_id: candidate)
+        retries += 1
+        raise "Failed to generate unique external_id after #{max_retries} attempts" if retries >= max_retries
+        candidate = SecureRandom.alphanumeric(12).downcase
+      end
+
+      candidate
+    end
```
Comment (EmCousin) [inline app/models/audience_export.rb:31]: This whole `external_id` logic is never used and would be overengineered for what we need here

File: app/models/audience_export.rb [hunk 9]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+class AudienceExport < ApplicationRecord
+  belongs_to :seller, class_name: "User"
+  belongs_to :recipient, class_name: "User"
+  has_many :chunks, class_name: "AudienceExportChunk", foreign_key: :export_id, dependent: :delete_all
+
+  serialize :audience_options, type: Hash, coder: YAML
```
Comment (EmCousin) [inline app/models/audience_export.rb:8]: Use `attr_json_data_accessor` from the `JsonData` module

File: app/models/audience_export.rb [hunk 10]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+class AudienceExport < ApplicationRecord
+  belongs_to :seller, class_name: "User"
+  belongs_to :recipient, class_name: "User"
+  has_many :chunks, class_name: "AudienceExportChunk", foreign_key: :export_id, dependent: :delete_all
```
Comment (EmCousin) [inline app/models/audience_export.rb:6]: Just rename the foreign key to `audience_export_id` and rely on Rails' conventions. Right now there is useless configuration and it does not make the code clearer. In the end, you'll be able to simplify:
```suggestion
  has_many :audience_export_chunks, dependent: :delete_all
```

File: app/sidekiq/exports/audience/compile_chunks_job.rb [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    @export = AudienceExport.find(export_id)
+
+    tempfile = generate_csv
+    send_email(tempfile)
+    cleanup
+  end
+
+  private
+    def generate_csv
+      timestamp = Time.current.to_fs(:db).gsub(/ |:/, "-")
```
Comment (EmCousin) [inline app/sidekiq/exports/audience/compile_chunks_job.rb:21]: Is there a better formatter so you don't have to `gsub` the formatted string?

File: app/sidekiq/exports/audience/create_and_enqueue_chunks_job.rb [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+      member_ids = build_query.order(:id).pluck(:id)
+
+      if member_ids.empty?
+        Exports::Audience::CompileChunksJob.perform_async(@export.id)
+        return
+      end
+
+      member_ids.each_slice(CHUNK_SIZE) do |ids_batch|
+        @export.chunks.create!(member_ids: ids_batch)
+      end
```
Comment (EmCousin) [inline app/sidekiq/exports/audience/create_and_enqueue_chunks_job.rb:29]: ⚠️ Careful here, you are requesting all the member ids at once before counting them in memory.
You can count all the members first in the database, then you can query them by chunk to make it more efficient


```suggestion
      if build_query.none?
        Exports::Audience::CompileChunksJob.perform_async(@export.id)
        return
      end

      build_query.select(:id).order(:id).in_batches(of: CHUNK_SIZE) do |members|
        @export.chunks.create!(member_ids: members.ids)
      end
```

File: app/sidekiq/exports/audience/create_and_enqueue_chunks_job.rb [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    end
+
+    def build_query
+      query = @export.seller.audience_members.select(:id)
+      conditions = []
+      options = @export.audience_options.with_indifferent_access
+      conditions << "follower = true" if options[:followers]
+      conditions << "customer = true" if options[:customers]
+      conditions << "affiliate = true" if options[:affiliates]
+      query.where(conditions.join(" OR "))
```
Comment (EmCousin) [inline app/sidekiq/exports/audience/create_and_enqueue_chunks_job.rb:44]:
```suggestion
    query = @export.seller.audience_members
    options = @export.audience_options.with_indifferent_access # is the indifferent access needed?
    
    scopes = []
    scopes << query.where(follower: true) if options[:followers]
    scopes << query.where(customer: true) if options[:customers]
    scopes << query.where(affiliate: true) if options[:affiliates]

    scopes.any? ? scopes.reduce(&:or) : query
```

File: app/sidekiq/exports/audience_export_worker.rb [hunk 14]
```diff
@@ -4,16 +4,47 @@ class Exports::AudienceExportWorker
...
   end
+
+  private
+    def audience_count
+      query = @seller.audience_members
+      conditions = []
+      conditions << "follower = true" if @audience_options[:followers]
+      conditions << "customer = true" if @audience_options[:customers]
+      conditions << "affiliate = true" if @audience_options[:affiliates]
+      query.where(conditions.join(" OR ")).count
```
Comment (EmCousin) [inline app/sidekiq/exports/audience_export_worker.rb:28]:
```suggestion
    query = @seller.audience_members
    
    scopes = []
    scopes << query.where(follower: true) if @audience_options[:followers]
    scopes << query.where(customer: true) if @audience_options[:customers]
    scopes << query.where(affiliate: true) if @audience_options[:affiliates]

    query = scopes.reduce(&:or) if scopes.any?
    query.count
```
